### DISCUSSION TITLE
Monitor connected MIDI devices?

### DISCUSSION THREAD

**Discussion by nejdu** [melanchall/drywetmidi#120] [General]
Thank you for a most excellent library! 

I'm not sure if this is possible or not, but is there any way to monitor if a MIDI device has disconnected? I'm thinking of simply running GetAll() every few seconds to see if the device list has changed, but that doesn't really feel great. My use case is that if someone trips over the MIDI keyboard cord I want to disconnect from the port and make it possible to re-connect without having to force a restart of everything.

**Comment by melanchall** [comment]
Hi,

Thanks for using the library!

Unfortunately it's not possible to catch MIDI device plugging/unplugging in Windows. Yes, you can call GetAll every second, but that's not a solution. What if two devices have the same name? When one is unplugged and another one is plugged, you're not able to determine what device plugged and what unplugged. With your algorithm you will think the first device is plugged again, but in fact the second one is plugged.

**Reply by nejdu** [reply]
Dang! I was afraid that was the case... I guess the only sure way is to provide a connect/disconnect button to manually manage the connection. Just to test, I tried connecting and disconnecting a Keyboard as MIDI input to the famous Kontakt sampler and it didn't detect the disconnection or tried to re-connect either, I had to restart for it to attach to the keyboard again. Oh well, I really hope Microsoft improves the Windows MIDI API with this some day! :)
